probability/windowed_median.cpp

```diff
@@ -0,0 +1,185 @@
+     */
+    void eraseFromSorted(int value) {
+        const auto sz = _sortedValues.size();
+        
+        // If the erased value is on the left branch or the median itself and the number of elements is even, the new
+        // median will be the right child of the current one
+        if (value <= *_itMedian && sz % 2 == 0) {
+            ++_itMedian;    // O(1) - traversing one step to the right child
+        }
+        
+        // However, is the erased value is on the right branch or the median itself, and the number of elements is odd,
+        // the new median will be the left child of the current one
+        else if (value >= *_itMedian && sz % 2 != 0) {
+            --_itMedian;    // O(1) - traversing one step to the left child
+        }
+        
+        // Find the (first) position of the value we want to erase, and erase it
+        const auto it = _sortedValues.find(value);  // O(logN)
+        _sortedValues.erase(it);                    // O(logN)
+    }
+    
+public:
+    
+    /**
+     * @brief Constructs a WindowedMedian object
+     * @param windowSize Sliding window size
+     */
+    explicit WindowedMedian(int windowSize) : _windowSize(windowSize) {};
+    
+    /**
+     * @brief Insert a new value to the stream
+     * @param value New value to insert
+     */
+    void insert(int value) {
+        
+        // Push new value to the back of the sliding window - O(1)
+        _window.push_back(value);
+        insertToSorted(value);  // Insert value to the multi-value BST - O(logN)
+        if (_window.size() > _windowSize) { // If exceeding size of window, pop from its left side
+            eraseFromSorted(_window.front());   // Erase from the multi-value BST the window left side value
+            _window.pop_front();    // Pop the left side value from the window - O(1)
+        }
+    }
+    
+    /**
+     * @brief Gets the median of the values in the sliding window
+     * @return Median of sliding window. For even window size return the average between the two values in the middle
+     */
+    float getMedian() const {
+        if (_sortedValues.size() % 2 != 0) {
+            return *_itMedian;  // O(1)
+        }
+        return 0.5f * *_itMedian + 0.5f * *next(_itMedian);   // O(1)
+    }
+    
+    /**
+     * @brief A naive and inefficient method to obtain the median of the sliding window. Used for testing!
+     * @return Median of sliding window. For even window size return the average between the two values in the middle
+     */
+    float getMedianNaive() const {
+        auto window = _window;
+        window.sort();      // Sort window - O(NlogN)
+        auto median = *next(window.begin(), window.size() / 2); // Find value in the middle - O(N)
+        if (window.size() % 2 != 0) {
+            return median;
+        }
+        return 0.5f * median + 0.5f * *next(window.begin(), window.size() / 2 - 1);   // O(N)
+    }
+};
+}   // namespace probability
+
+#include <vector>
```

Comment:
Shouldn't this header be included above?

Reply:
I can move those includes to the file header. It's just that these specific includes are needed only by `test()` and `main()`. The core algorithmic class doesn't need those. So, it's a question of what you prefer as the standard. Should I include everything in the file header, or only libraries that are needed by the algorithm?

Reply:
OK